# LGT3109 Tutorial 4 Report (Sample Solutions)

#### Input your name and student ID in the cell below (<font color='red'>if a cell is not in edit mode, double click it</font>):

Your name: Dai Cijun

Your student ID:22102054d

#### **Instructions for the report**:
* Open Anaconda Navigator, and launch Python IDLE through Anaconda Navigation
* Refer to Section 3.1 of the tutorial book and open tutorial 4 report
* Follow Section 4 of the tutorial book to save the report and zip the report folder. The zip file is named as LGT3109T4Report{your student_id}.zip (e.g., if student_id is 12345678d, then the zip file's name is LGT3109T4Report12345678d.zip). The zip file needs to include the following files:
    - an .ipynb file of this tutorial report    
    - image files of screenshots used in this tutorial report  
* Submit the zip file of the report folder in the blackboard. The submission due date is **<font color='red'>30 Sep 2025</font>**

### Part 1 Python Function

#### 1.1. Read the following code and answer questions below (20 points):

In [36]:
def max_num(num1, num2):
    result = num1
    if num2>result:
        result = num2
    return result

q1 = int(input('Enter the shipping quantity of carrier 1: '))
q2 = int(input('Enter the shipping quantity of carrier 2: '))
largest_q = max_num(q1, q2)
print('The largest shipping quantity is:', largest_q)

The largest shipping quantity is: 30


Question 1.1: List all the functions used in the code above, explain what types of functions they are, and explain what they are used for. 

<font color='red'>*Edit this cell to list the functions and explain what they are used for:*</font>  
1. "max_number" function: User-defined function, used to find the bigger number between two numbers.
2. "int" function: Built-in function, used to convert a value to an integer.
3. "input" function: Built-in function, used to get user input.
4. "print" function: Built-in function, used to output the information.


Question 1.2: Revise the statement ``largest_q = max_num(q1, q2)``, so that given a shipping quantity of an additional 3rd carrier as part of the input, the code outputs the largest shipping quantity among the three shipping quantities. Based on this, explain the benefits brought by the use of a function.

In [ ]:
#revise the code below accordingly:
def max_num(num1, num2):
    result = num1
    if num2>result:
        result = num2
    return result

q1 = int(input('Enter the shipping quantity of carrier 1: '))
q2 = int(input('Enter the shipping quantity of carrier 2: '))
largest_q = max_num(q1, q2)
print('The largest shipping quantity is:', largest_q)

Enter the shipping quantity of carrier 1: 10
Enter the shipping quantity of carrier 2: 20
The largest shipping quantity is: 20


<font color='red'>*Edit this cell to explain the benefits brought by the use of a function, based on the revised code above:*</font>  
1.Reusability: Functions could be used multiple times without rewriting the code. (The max_number fuction could easily be used to compare three numbers by calling it twice.)  
2.Easy to read: Functions could make the code in shorter and easier to read.  
3. Consistence: Functions could avoid mistakes when handwriting the same code multiple times.



### Part 2 User-defined Functions

#### 2.1. Execute each of the codes below, explain the errors received, and discuss how to fix the errors (8 points):

##### (a)

In [37]:
def negative_num(num):
res = 0
if num<=0:
    res = num
return res

print(negative_num(10))
print(negative_num(-10))

IndentationError: expected an indented block after function definition on line 1 (4210116866.py, line 2)

<font color='red'>*Edit this cell to explain errors received from the code above, and discuss how to fix the errors:*</font>
There's a type error, because there should be a indent before "res = 0" to indicate that it's in the function body.


In [38]:
#Copy the corrected code below in this cell:
def negative_num(num):
    res = 0
    if num<=0:
        res = num
    return res
print(negative_num(10))
print(negative_num(-10))

0
-10


##### (b)

In [39]:
def negative_num(num):
    res = 0
    if num<=0:
        res = num
    return res

negative_num(-10)
print(res)

NameError: name 'res' is not defined

<font color='red'>*Edit this cell to explain errors received from the code above, and discuss how to fix the errors:*</font>  
"res" is the variable that only defined in the user-built function "negative_num", so it can't be used outside the fuction. To fix that, just write "print(negative_num(-10))" to print the return value of the function.

In [40]:
#Copy the corrected code below in this cell:
def negative_num(num):
    res = 0
    if num<=0:
        res = num
    return res

print(negative_num(-10))

-10


#### 2.2. Execute the codes below, and answer the questions (12 points):

##### (a)

In [41]:
def compute_shipping_cost(quantity,rate):
    cost = quantity * rate
    return cost

cost = compute_shipping_cost(10,100.5)
compute_shipping_cost(10,10)
print(cost)

1005.0


<font color='red'>*Edit this cell to explain why in the code above, after executing ``compute_shipping_cost(10,10)``, the value of variable ``cost`` is unchanged:*</font>  
Because "cost" is a variable that only defined in the user-built function "compute_shipping_cost", so it can't be used outside the function. The "cost" outside the function is a different variable, which is unchanged.



##### (b)

In [42]:
def compute_shipping_cost(quantity=10,rate=50):
    cost = quantity * rate
    return cost

cost1 = compute_shipping_cost(10,50)
cost2 = compute_shipping_cost(10)
cost3 = compute_shipping_cost()
print(cost1,cost2,cost3)

500 500 500


<font color='red'>*Edit this cell to explain why after the code is executed, ``cost1``, ``cost2``, and ``cost3`` have the same value:*</font>  
Because the function has already defined the value of quantity and rate, which are 10 and 50, so no matter what values are input when calling the function, the return value is always 500.



##### (c)

In [43]:
def compute_shipping_cost(quantity,rate):
    global cost
    cost = quantity * rate
    return cost

cost = 0
print (cost)
compute_shipping_cost(10,100.5)
print(cost)

0
1005.0


<font color='red'>*Edit this cell to explain why in the code above, the value of variable ``cost`` is changed after the call of function ``compute_shipping_cost(10,100.5)``:*</font>  

Because the function global cost would make the cost inside the function the same variable as the cost outside the function, so when the function is called, the value of cost would be changed to 1005.0.




### Part 3 Built-in Functions and Modules

#### 3.1. Use help() to show the definition and usage of function ``print``, and answer the following questions (10 points):

In [44]:
#edit this cell to write a code to use help() to show the definition and usage of function print:
help(print)


Help on built-in function print in module builtins:

print(*args, sep=' ', end='\n', file=None, flush=False)
    Prints the values to a stream, or to sys.stdout by default.

    sep
      string inserted between values, default a space.
    end
      string appended after the last value, default a newline.
    file
      a file-like object (stream); defaults to the current sys.stdout.
    flush
      whether to forcibly flush the stream.



Question 3.1.1. Explain which parameters of function ``print`` have default values? What are these default values?

<font color='red'>*Edit this cell to answer the question above:*</font>  
The parameters that have default values are sep, end, file and flush. Their default values are ' ' (a space), '\n' (a new line), sys.stdout and False respectively.


Question 3.1.2. Given three string variables ``s1``, ``s2``, and ``s3``, explain how to use function ``print`` and use keyword arguments to output a line that concatenates the three strings with ', ' instead of ' ' between the strings. For example, if s1='Apple', s2='Orange', and s3='Grapes', then it outputs a line of 'Apple, Orange, Grapes'

In [45]:
#edit this cell to write a code for the question above  
s1 = 'Apple'
s2 = 'Orange'
s3 = 'Grapes'
print(s1, s2, s3, sep=', ')


Apple, Orange, Grapes


### Part 4 Other Exercises

#### 4.1. (Promotion Letter) Company FE would like to generate a letter to send a promotion code to its customer (25 points). 
The letter is as follows:

    Dear {Title} {Customer Name},
        Thank you very much for ordering our product. 
        Your promotion code is {Promotion Code}.
    Best,
    {Your Name}

Here, {Title}, {Customer Name}, and {Promotion Code} are from the input, and {Your Name} is your name.

Write a function to generate the letter for given information about {Title}, {Customer Name}, {Promotion Code}, and {Your Name}, where {Title}, {Customer Name}, and {Promotion Code} are from the input, and and {Your Name} is fixed to be your actual name (e.g., for me, {Your Name}='Xiaoyu'). 

##### (a)

In [46]:
#edit this cell to define the function and write its body:
def letter_generate(Title, Customer_name, Promotion_code, Your_name='Dai Cijun'):
    letter = (
        f"Dear {Title} {Customer_name},\n"
        f"    Thank you very much for ordering our product. \n"
        f"    Your promotion code is {Promotion_code}.\n"
        f"Best,\n"
        f"{Your_name}"
    )
    return letter


##### (b)

In [47]:
#edit this cell to write a code that uses the function above to generate three letters, 
# where {Title} and {Customer Name} are from the input, 
#and {Your Name} is fixed to be your actual name (e.g., for me, {Your Name}='Xiaoyu')
Title = input('enter Title:')
Customer_name = input('enter Customer Name:')
Promotion_code = input('enter Promotion Code:')
letter =letter_generate(Title, Customer_name, Promotion_code)
print(letter)

Dear PolyU Mr.Wang,
    Thank you very much for ordering our product. 
    Your promotion code is 20250923.
Best,
Dai Cijun


##### (c)

<font color='red'>Execute and test your revised code above for various inputs. Edit this cell to copy your test results:</font>  
Dear PolyU Mr.Wang,    
    Thank you very much for ordering our product.     
    Your promotion code is 20250923.  
Best,  
Dai Cijun



#### 4.2. (Event Record) Write a function to print a record of an event in the following format (25 points):
    {Event Name}    
    Location: {Location}                   
    Time: {Time}
    Participants: {Participants}
    
where {Event Name}, {Location}, {Time}, and {Participants} are four parameters of the function.

For example, for an event of my weekly research meeting with students, the record printed is as follows:

    Weekly Research Meeting    
    Location: M505e
    Time: 10:00, Every Friday
    Participants: Professor Xiaoyu Wang and his students

##### (a)

In [3]:
#Edit this cell to define the function and write its body:
def event_generate(Event_name, Location, Time, Participants):
    event = (
        f"{Event_name},\n"
        f"Location:{Location},\n"
        f"Time:{Time},\n"
        f"Participants:{Participants}"
    )
    return event


##### (b)

In [4]:
#Edit this cell to write a statement to use the function above to print the event of my weekly research meeting:
event1 = event_generate("Weekly Research Meeting", "M505e", "10:00, Every Friday", "Professor Xiaoyu Wang and his students")
print(event1)


Weekly Research Meeting,
Location:M505e,
Time:10:00, Every Friday,
Participants:Professor Xiaoyu Wang and his students


##### (c)

In [50]:
#Edit this cell to write a statement to use the function above to print an event of which you are one of the participants.
event2 = event_generate("LGT3109 Course", "MN102C", "08:30. Every Tuesday", "Dai Cijun")
print(event2)

LGT3109 Course,
Location:MN102C,
Time:08:30. Every Tuesday,
Participants:Dai Cijun

